analytics/uptime_data.go

```diff
@@ -1,20 +1,332 @@
+}
+
+type UptimeReportAggregateSQL struct {
+	ID string `gorm:"primaryKey"`
+
+	Counter `json:"counter" gorm:"embedded"`
+
+	TimeStamp      int64  `json:"timestamp" gorm:"index:dimension, priority:1"`
+	OrgID          string `json:"org_id" gorm:"index:dimension, priority:2"`
+	Dimension      string `json:"dimension" gorm:"index:dimension, priority:3"`
+	DimensionValue string `json:"dimension_value" gorm:"index:dimension, priority:4"`
+
+	Code1x  int `json:"code_1x"`
+	Code200 int `json:"code_200"`
+	Code201 int `json:"code_201"`
+	Code2x  int `json:"code_2x"`
+	Code301 int `json:"code_301"`
+	Code302 int `json:"code_302"`
+	Code303 int `json:"code_303"`
+	Code304 int `json:"code_304"`
+	Code3x  int `json:"code_3x"`
+	Code400 int `json:"code_400"`
+	Code401 int `json:"code_401"`
+	Code403 int `json:"code_403"`
+	Code404 int `json:"code_404"`
+	Code429 int `json:"code_429"`
+	Code4x  int `json:"code_4x"`
+	Code500 int `json:"code_500"`
+	Code501 int `json:"code_501"`
+	Code502 int `json:"code_502"`
+	Code503 int `json:"code_503"`
+	Code504 int `json:"code_504"`
+	Code5x  int `json:"code_5x"`
+}
+
+func (a UptimeReportAggregateSQL) GetAssignments(tableName string) map[string]interface{} {
+	assignments := make(map[string]interface{})
+
+	baseFields := structs.Fields(a)
+	for _, field := range baseFields {
+		colName := field.Tag("json")
+		if strings.Contains(colName, "code_") {
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr(tableName + "." + colName + " + " + fmt.Sprint(field.Value()))
+			}
+		}
+	}
+
+	fields := structs.Fields(a.Counter)
+	for _, field := range fields {
+		colName := "counter_" + field.Tag("json")
+
+		switch {
+		case strings.Contains(colName, "hits"), strings.Contains(colName, "error"), strings.Contains(colName, "success"):
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr(tableName + "." + colName + " + " + fmt.Sprint(field.Value()))
+			}
+		case strings.Contains(colName, "total_request_time"):
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr(tableName + "." + colName + " + " + fmt.Sprint(a.TotalRequestTime))
+			}
+		case strings.Contains(colName, "request_time"):
+			//AVG adding value to another AVG: newAve = ((oldAve*oldNumPoints) + x)/(oldNumPoints+1)
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr("((" + tableName + "." + colName + " * " + tableName + ".counter_hits) + " + fmt.Sprint(a.RequestTime) + ")/( " + tableName + ".counter_hits +1)")
+			}
+		}
+
+	}
+
+	return assignments
+}
+
+func (u *UptimeReportAggregate) Dimensions() (dimensions []Dimension) {
+	for key, inc := range u.URL {
+		dimensions = append(dimensions, Dimension{"url", key, inc})
+	}
+
+	for key, inc := range u.Errors {
+		dimensions = append(dimensions, Dimension{"errors", key, inc})
+	}
+
+	dimensions = append(dimensions, Dimension{"", "total", &u.Total})
+
+	return
+}
+
+func (a *UptimeReportAggregateSQL) TableName() string {
+	return "tyk_uptime_analytics"
```

Comment:
We are using this `tyk_uptime_analytics` throughout the PR, we can put it as a `const`

Reply:
Done